src-java/flowhs-topology/flowhs-storm-topology/src/test/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationTestBase.java

```diff
@@ -0,0 +1,505 @@
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+
+package org.openkilda.wfm.topology.flowhs.fsm.haflow.validation;
+
+import static org.openkilda.rulemanager.OfVersion.OF_13;
+
+import org.openkilda.messaging.info.flow.FlowDumpResponse;
+import org.openkilda.messaging.info.group.GroupDumpResponse;
+import org.openkilda.messaging.info.meter.MeterDumpResponse;
+import org.openkilda.model.FlowPathDirection;
+import org.openkilda.model.GroupId;
+import org.openkilda.model.IpSocketAddress;
+import org.openkilda.model.Meter;
+import org.openkilda.model.MeterId;
+import org.openkilda.model.Switch;
+import org.openkilda.model.SwitchFeature;
+import org.openkilda.model.SwitchId;
+import org.openkilda.model.SwitchStatus;
+import org.openkilda.model.cookie.Cookie;
+import org.openkilda.model.cookie.FlowSegmentCookie;
+import org.openkilda.persistence.inmemory.InMemoryGraphBasedTest;
+import org.openkilda.persistence.repositories.FlowPathRepository;
+import org.openkilda.persistence.repositories.HaFlowPathRepository;
+import org.openkilda.persistence.repositories.HaFlowRepository;
+import org.openkilda.persistence.repositories.HaSubFlowRepository;
+import org.openkilda.persistence.repositories.SwitchRepository;
+import org.openkilda.persistence.repositories.TransitVlanRepository;
+import org.openkilda.persistence.repositories.VxlanRepository;
+import org.openkilda.rulemanager.Field;
+import org.openkilda.rulemanager.FlowSpeakerData;
+import org.openkilda.rulemanager.GroupSpeakerData;
+import org.openkilda.rulemanager.Instructions;
+import org.openkilda.rulemanager.MeterFlag;
+import org.openkilda.rulemanager.MeterSpeakerData;
+import org.openkilda.rulemanager.OfFlowFlag;
+import org.openkilda.rulemanager.OfTable;
+import org.openkilda.rulemanager.ProtoConstants.PortNumber;
+import org.openkilda.rulemanager.SpeakerData;
+import org.openkilda.rulemanager.action.Action;
+import org.openkilda.rulemanager.action.GroupAction;
+import org.openkilda.rulemanager.action.PopVlanAction;
+import org.openkilda.rulemanager.action.PortOutAction;
+import org.openkilda.rulemanager.action.PushVlanAction;
+import org.openkilda.rulemanager.action.SetFieldAction;
+import org.openkilda.rulemanager.group.Bucket;
+import org.openkilda.rulemanager.group.GroupType;
+import org.openkilda.rulemanager.group.WatchGroup;
+import org.openkilda.rulemanager.group.WatchPort;
+import org.openkilda.rulemanager.match.FieldMatch;
+import org.openkilda.wfm.share.flow.resources.FlowResourcesConfig;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Sets;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class HaFlowValidationTestBase extends InMemoryGraphBasedTest {
```

Comment:
not for this review, but in general we need to revisit the usage of test base classes. Typically, reusability of the code is worse when using inheritance, there's no IS-A relationship. "Prefer composition over inheritance" probably needs to be applied here. 
The modification would be to change these base classes to distinct service classes that we'll inject as member variables. Creating switches and other data seems reusable, but it's difficult to reuse, because another test would extend another base class and can't benefit from code in this class.

Reply:
great point, I removed it since we do not need it here.